### PR TITLE
Add product management SKU

### LINKED ISSUE #34386: [Inventory] SKU field
**User story**

As a merchant,
I want to control my product inventory
So that I can keep up with the demand

As a merchant,
I want to manage my inventory
So that my customers won’t purchase products that I don’t have at hand

**Design**
5sAIeTRd9Yp7nSCT33BAWz-fi-6498%3A284871

**Acceptance criteria**

- [ ] Created automatically by abbreviating the product name.
- [ ] Accepts letters and numbers.
- [ ] 160 character limit (currently unlimited, but we’ve got placed in the new UI where it’ll appear and need to introduce some constraints to guarantee legibility).

### PR DESCRIPTION
**Pull request by joshuatf**
### All Submissions:

-   [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change?

### Changes proposed in this Pull Request:

Adds the SKU field and inventory section to the new product management experience.

Closes #34386 .
Screen Shot 2022-10-06 at 7 12 24 AM
Screen Shot 2022-10-06 at 7 12 10 AM
Screen Shot 2022-10-06 at 7 11 59 AM

### How to test the changes in this Pull Request:

1. Visit the new product page `Products -> Add new (MVP)`
2. Type in a product name and blur the input
3. Note the SKU has been added in slug format to the SKU field
4. Change the SKU to a bad format including non-alphanumeric characters and spaces
5. Note the error
6. Make a very long SKU (greater than 160 characters) and note the error
7. Save and make sure the SKU persists on refresh
8. Remove the product name so it's empty
9. Blur the field and make sure the error appears

### Other information:

-   [x] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [x] Have you written new tests for your changes, as applicable?
-   [x] Have you successfully run tests with your changes locally?
-   [x] Have you created a changelog file for each project being changed, ie `pnpm --filter=<project> run changelog add`?

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

### CODE REVIEW

File: plugins/woocommerce-admin/client/products/product-validation.ts [hunk 1]
```diff
@@ -51,6 +56,20 @@ export const validate = (
 		);
 	}
 
+	if ( values.sku && values.sku.length > 0 && ! isValidSlug( values.sku ) ) {
```
Comment (mdperez86) [inline plugins/woocommerce-admin/client/products/product-validation.ts]: I think we can group these 2 conditions like
```javascript
if ( values.sku ) {
    if ( values.sku.length > 0 && ! isValidSlug( values.sku ) ) { ... }
    if ( values.sku.length > 160 ) { ... }
}
```

File: plugins/woocommerce-admin/client/products/sections/product-details-section.tsx [hunk 2]
```diff
@@ -74,33 +81,34 @@ export const ProductDetailsSection: React.FC = () => {
 							{ ...getTextControlProps(
 								getInputProps( 'name' )
 							) }
+							onBlur={ () => {
+								setSkuIfEmpty();
```
Comment (mdperez86) [inline plugins/woocommerce-admin/client/products/sections/product-details-section.tsx:85]: If you override the onBlur then you should call the parent onBlur manualy like:
```javascript
const nameTextControlProps = getTextControlProps( getInputProps( 'name' ) );
...
onBlur={ () => {.
    ...
    nameTextControlProps?.onBlur();
} }
```
Validations and other internal form states change when onBlur happened.

**Review by mdperez86 — CHANGES_REQUESTED** [review]

File: plugins/woocommerce-admin/client/products/sections/product-details-section.tsx (hunk 2 above)
Comment (joshuatf) [reply plugins/woocommerce-admin/client/products/sections/product-details-section.tsx:85]: Good catch, @mdperez86!  Updated in https://github.com/woocommerce/woocommerce/pull/34978/commits/b00dbf96c6e29e6cc3902db89a32bb5ff01fc2d2.

On a side note, I think the `getTextControl` and other wrappers around `getInputProps` are relatively thin at the moment.  I think it's also worth implementing some of this logic in the `Form` component to prevent all of these parent calls that we're going to see throughout this experience.

File: plugins/woocommerce-admin/client/products/product-validation.ts (hunk 1 above)
Comment (joshuatf) [reply plugins/woocommerce-admin/client/products/product-validation.ts]: In general, we try to avoid nesting like this.  It only shortens the line by a few characters but reduces readability.  The same applies for our SCSS selectors where we try not to go more than 3 levels deep (though this is not a firm rule).

**Comment by joshuatf** [comment]
Thanks for the review, Maikel!  This PR is ready for another look when you have time.

File: plugins/woocommerce-admin/client/products/product-validation.ts [hunk 3]
```diff
@@ -10,6 +10,11 @@ import {
 } from '@woocommerce/data';
 import type { FormErrors } from '@woocommerce/components';
 
+/**
+ * Internal dependencies
+ */
+import { isValidSlug } from './sections/utils';
```
Comment (mattsherman) [inline plugins/woocommerce-admin/client/products/product-validation.ts]: Instead of implementing `isValidSlug` in utils (which feels like a junk drawer), what about implementing it directly in this `product-validation.ts` file and exporting it from here. Feels like a better place for it, and still allows re-use.

File: plugins/woocommerce-admin/client/products/sections/product-details-section.tsx (hunk 2 above)
Comment (joshuatf) [reply plugins/woocommerce-admin/client/products/sections/product-details-section.tsx:85]: Issue created to tackle the `getInputProps` changes made [here](https://github.com/woocommerce/woocommerce/issues/34996).

File: plugins/woocommerce-admin/client/products/product-validation.ts (hunk 1 above)
Comment (mattsherman) [reply plugins/woocommerce-admin/client/products/product-validation.ts]: You could use `values.sku?.length`.

File: plugins/woocommerce-admin/client/products/product-validation.ts [hunk 4]
```diff
@@ -51,6 +56,20 @@ export const validate = (
...
 	}
 
+	if ( values.sku && values.sku.length > 0 && ! isValidSlug( values.sku ) ) {
+		errors.sku = __(
+			'SKU can only contain alphanumeric characters, hyphens, and underscores.',
+			'woocommerce'
+		);
+	}
+
+	if ( values.sku && values.sku.length > 160 ) {
```
Comment (mattsherman) [inline plugins/woocommerce-admin/client/products/product-validation.ts]: I missed that we were changing the SKU validation from what currently exists in core in the original issue. I don't think we should change the validation. Otherwise, merchants will not be able to edit existing products created outside of the new experience (whether via the old experience or by the API) in the new experience.

From https://woocommerce.com/document/managing-products/....

> The SKU field can accept up to 255 characters, which can include alphanumeric values and some special characters. The input is checked using the WordPress function santize_text_field, which means HTML, invalid UTF characters, and octets will be removed when saved (an octet is defined by % followed by 2 characters with a value in the range of a-f or 0-9).

/cc @jarekmorawski @pmcpinto

File: plugins/woocommerce-admin/client/products/sections/product-details-section.tsx [hunk 5]
```diff
@@ -49,9 +49,16 @@ export const ProductDetailsSection: React.FC = () => {
 		}
 	);
 
-	function doesNameHaveError(): boolean {
+	const hasNameError = () => {
```
Comment (mattsherman) [inline plugins/woocommerce-admin/client/products/sections/product-details-section.tsx:52]: This name change seems unrelated to this PR.

File: plugins/woocommerce-admin/client/products/sections/product-details-section.tsx [hunk 6]
```diff
@@ -74,33 +81,35 @@ export const ProductDetailsSection: React.FC = () => {
 							{ ...getTextControlProps(
 								getInputProps( 'name' )
 							) }
+							onBlur={ () => {
```
Comment (mattsherman) [inline plugins/woocommerce-admin/client/products/sections/product-details-section.tsx:84]: Instead of tying this to the name field's blur, could we hook things up to set the SKU if empty when the underlying name data changes?

File: plugins/woocommerce-admin/client/products/sections/product-inventory-section.tsx [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+						) }
+					</span>
+					<Link
+						href={ getAdminLink(
+							'admin.php?page=wc-settings&tab=products&section=inventory'
+						) }
+						target="_blank"
+						type="wp-admin"
+						onClick={ () => {
+							recordEvent( 'add_product_inventory_help' );
```
Comment (mattsherman) [inline plugins/woocommerce-admin/client/products/sections/product-inventory-section.tsx:38]: I think the Tracks event should be something like `product_editor_inventory_help_clicked`.

* `add_product` isn't accurate -- this will also be fired when editing a product
* There is no action

(naming guidelines, specifying that an action should be in the name: PCYsg-4S2-p2)

File: plugins/woocommerce-admin/client/products/sections/test/utils.ts [hunk 8]
```diff
@@ -0,0 +1,23 @@
+/**
+ * Internal dependencies
+ */
+import { isValidSlug } from '../utils';
+
+describe( 'product section utils', () => {
+	describe( 'isValidSlug', () => {
```
Comment (mattsherman) [inline plugins/woocommerce-admin/client/products/sections/test/utils.ts]: As mentioned above, I think that we should respect the current validation constraints in core.

File: plugins/woocommerce-admin/client/products/sections/test/utils.ts [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+		it( 'should return true if a slug is valid', () => {
+			const validity = isValidSlug( 'valid-slug_123' );
+			expect( validity ).toBeTruthy();
```
Comment (mattsherman) [inline plugins/woocommerce-admin/client/products/sections/test/utils.ts]: I think `.toBeTrue()` would be better here, since we are in control of `isValidSlug()` and we really should only ever be returing `true` or `false`. It we were to change the implementation (by mistake) of `isValidSlug()` and it returned something like the slug itself, we'd want to catch that. `.toBeTruthy()` would still pass in that case.

File: plugins/woocommerce-admin/client/products/sections/test/utils.ts [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+		} );
+
+		it( 'should return false if a slug contains spaces', () => {
+			const validity = isValidSlug( 'Invalid slug' );
+			expect( validity ).toBeFalsy();
```
Comment (mattsherman) [inline plugins/woocommerce-admin/client/products/sections/test/utils.ts]: Should be `.toBeFalse()`

File: plugins/woocommerce-admin/client/products/sections/test/utils.ts [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+		} );
+
+		it( 'should return false if invalid characters exist', () => {
+			const validity = isValidSlug( 'invalid-characters!' );
+			expect( validity ).toBeFalsy();
```
Comment (mattsherman) [inline plugins/woocommerce-admin/client/products/sections/test/utils.ts]: Should be `.toBeFalse()`

File: plugins/woocommerce-admin/client/products/sections/utils.ts [hunk 12]
```diff
@@ -108,3 +108,13 @@ export const getInputControlProps = ( {
...
 	};
 };
+
+/**
+ * Test if a string is a valid slug.
+ *
+ * @param  string String to test
+ * @return boolean
+ */
+export const isValidSlug = ( string: string ) => {
```
Comment (mattsherman) [inline plugins/woocommerce-admin/client/products/sections/utils.ts]: Shouldn't this also include a length check? It isn't really a valid slug if it is greater than 255 characters (core validation).

File: plugins/woocommerce-admin/client/products/product-validation.ts (hunk 4 above)
Comment (mattsherman) [reply plugins/woocommerce-admin/client/products/product-validation.ts]: Also, shouldn't we be checking if it is a valid `SKU`, not a valid `slug`?

File: plugins/woocommerce-admin/client/products/sections/utils.ts (hunk 12 above)
Comment (mattsherman) [inline plugins/woocommerce-admin/client/products/sections/utils.ts]: Shouldn't this be `isValidSKU`?

File: plugins/woocommerce-admin/client/products/product-validation.ts (hunk 1 above)
Comment (mattsherman) [reply plugins/woocommerce-admin/client/products/product-validation.ts]: Shouldn't we check if it is a valid SKU, not a valid slug? Even if the current implementation has those being the exact same checks, it would be good to make an explicit check whether it is a valid SKU, since then in the future, if the validation of the two differed, this code would still be correct.

**Review by mattsherman — CHANGES_REQUESTED** [review]
Nice work on this so far, @joshuatf !

My comments mainly revolve around validation.

I totally missed that we were proposing changing the validation rules for SKU in the original issue. I don't think we should do so, since it will introduce compatibility issues with products added via other means (existing UI, API). /cc @pmcpinto @jarekmorawski

File: plugins/woocommerce-admin/client/products/product-validation.ts (hunk 3 above)
Comment (joshuatf) [reply plugins/woocommerce-admin/client/products/product-validation.ts]: I was actually thinking this should be the opposite and elevated up to a more global utils place to use since we'll end up using this in more places soon.

File: plugins/woocommerce-admin/client/products/product-validation.ts (hunk 1 above)
Comment (joshuatf) [reply plugins/woocommerce-admin/client/products/product-validation.ts]: >You could use values.sku?.length.

Good call!  I'll add this.

>Shouldn't we check if it is a valid SKU, not a valid slug

I'm wary of doing this because there is no such thing as a valid SKU in the eyes of WooCommerce.  We've changed the behavior here and we are in fact checking against the "slug rules."  

Should those two rules separate in the future, I think it would be worth renaming, but I think it would be dishonest to name this `isValidSku` when that does not have its own set of rules yet.

File: plugins/woocommerce-admin/client/products/product-validation.ts (hunk 4 above)
Comment (joshuatf) [reply plugins/woocommerce-admin/client/products/product-validation.ts]: >I don't think we should change the validation. Otherwise, merchants will not be able to edit existing products created outside of the new experience

That's a great point.  I think that alone is enough reason to remove the validation and sanitization from this.  Though we could keep the initial "slug" version on new products.

>Also, shouldn't we be checking if it is a valid SKU, not a valid slug?

Responded to this above, but as you noted, there are not SKU validation rules in core.  We are checking for a valid slug here.

File: plugins/woocommerce-admin/client/products/sections/product-details-section.tsx (hunk 5 above)
Comment (joshuatf) [reply plugins/woocommerce-admin/client/products/sections/product-details-section.tsx:52]: No it's not.  This is a general clean-up to match our naming conventions.  Do you think this warrants a separate PR?

File: plugins/woocommerce-admin/client/products/sections/product-details-section.tsx (hunk 6 above)
Comment (joshuatf) [reply plugins/woocommerce-admin/client/products/sections/product-details-section.tsx:84]: Great idea!  I'll give this a go.

File: plugins/woocommerce-admin/client/products/sections/product-inventory-section.tsx (hunk 7 above)
Comment (joshuatf) [reply plugins/woocommerce-admin/client/products/sections/product-inventory-section.tsx:38]: Agreed.  I was following the existing naming conventions that were already in place for the product experience.  Is there a sheet somewhere specifying those names or were these created on the spot?  cc @pmcpinto @octaedro

File: plugins/woocommerce-admin/client/products/sections/utils.ts (hunk 12 above)
Comment (joshuatf) [reply plugins/woocommerce-admin/client/products/sections/utils.ts]: Same comment as above.

Comment (joshuatf) [reply plugins/woocommerce-admin/client/products/sections/utils.ts]: I think you're right.  Interestingly, `@wordpress/url`'s `cleanSlug` does not take into account length, but that may be by design while validation is meant to be separate.

I'll add the length logic here if we end up keeping the validation.

**Comment by pmcpinto** [comment]
> I totally missed that we were proposing changing the validation rules for SKU in the original issue. I don't think we should do so, since it will introduce compatibility issues with products added via other means (existing UI, API). /cc @pmcpinto @jarekmorawski

Good point. I agree with @mattsherman assessment. IMO the benefits of this change don't outweigh the associated risks

File: plugins/woocommerce-admin/client/products/product-validation.ts (hunk 1 above)
Comment (mdperez86) [reply plugins/woocommerce-admin/client/products/product-validation.ts]: > In general, we try to avoid nesting like this. It only shortens the line by a few characters but reduces readability. The same applies for our SCSS selectors where we try not to go more than 3 levels deep (though this is not a firm rule).

No it's not a firm rule, it's a fact! I agree excessive nesting can be hard to read and understand but nesting by a context with not to much levels helps a lot to find and understand things. For example in this case we are talking about all validations that are related to the sku.

File: plugins/woocommerce-admin/client/products/sections/product-details-section.tsx (hunk 2 above)
Comment (mdperez86) [reply plugins/woocommerce-admin/client/products/sections/product-details-section.tsx:85]: > getTextControlProps( getInputProps( 'name' ) )

I don't like this. I prefer a Hoc or a [controller](https://react-hook-form.com/get-started/#IntegratingControlledInputs) component to adapt props. So the `getInputProps` or what ever the name should return all necesary props related to the form logic and not to the control we are registering to the form logic. 

> I think it's also worth implementing some of this logic in the Form component to prevent all of these parent calls that we're going to see throughout this experience.

Same page with you

Comment (mdperez86) [reply plugins/woocommerce-admin/client/products/sections/product-details-section.tsx:85]: > Issue created to tackle the getInputProps changes made [here](https://github.com/woocommerce/woocommerce/issues/34996).

Awesome!!!!

**Review by mdperez86 — DISMISSED** [review]

File: plugins/woocommerce-admin/client/products/product-validation.ts (hunk 3 above)
Comment (mattsherman) [reply plugins/woocommerce-admin/client/products/product-validation.ts]: I'm not a fan of utils, in general. They always feel like a "junk drawer" for functions.

File: plugins/woocommerce-admin/client/products/sections/product-details-section.tsx (hunk 5 above)
Comment (mattsherman) [reply plugins/woocommerce-admin/client/products/sections/product-details-section.tsx:52]: I ultimately don't think that this file should need to be modified to implement the SKU (re: my comment about `onBlur` below). So, if that is the case, then changing this function name in the context of this PR is unrelated. 

However, if this file needs to be modified, since this function is local to the file, I think it seems like a safe clean-up to make in the context of this PR, though still unnecessary and it does add a little more that should be tested/reviewed in the context of this PR (to make sure that name error handling still works, but honestly that is largely a compilation issue since we are in TS here). So, I'm cool with it!

File: plugins/woocommerce-admin/client/products/sections/utils.ts (hunk 12 above)
Comment (mattsherman) [reply plugins/woocommerce-admin/client/products/sections/utils.ts]: Yeah, the concepts of cleaning/sanitization and validation are overlapping but different. When talking about whether something is "valid", I think we are talking about validation, which I think is a superset that includes cleaning/sanitization and additional rules.

File: plugins/woocommerce-admin/client/products/sections/product-details-section.tsx (hunk 6 above)
Comment (mattsherman) [reply plugins/woocommerce-admin/client/products/sections/product-details-section.tsx:84]: Sweet! I really think that is the way to go, as it keeps the different sections more decoupled and is also how ideally extensions would hook into things.

File: plugins/woocommerce-admin/client/products/product-validation.ts (hunk 4 above)
Comment (joshuatf) [reply plugins/woocommerce-admin/client/products/product-validation.ts]: All sku validation has been removed from this PR to align with core.

File: plugins/woocommerce-admin/client/products/sections/product-details-section.tsx (hunk 5 above)
Comment (joshuatf) [reply plugins/woocommerce-admin/client/products/sections/product-details-section.tsx:52]: I've left this in the PR.  Worth double checking the product name validation by focusing the input, emptying the field, and blurring the input to make sure the error still appears.  I'll add to the above testing instructions.
Screen Shot 2022-10-10 at 12 03 37 PM

File: plugins/woocommerce-admin/client/products/sections/product-details-section.tsx (hunk 6 above)
Comment (joshuatf) [reply plugins/woocommerce-admin/client/products/sections/product-details-section.tsx:84]: There's not a very straightforward to do this since watching `values` returns changes on each character addition.  I opted to watch for previous touched state and compare to the current touched state of "name."

While this does not give us a 1 to 1 for blur events, it performs reasonably similarly and should accomplish the criteria needed for this PR.

https://github.com/woocommerce/woocommerce/pull/34978/commits/d08c11b47a09e24a7f632d2f71ed4682b75d342c

File: plugins/woocommerce-admin/client/products/sections/test/utils.ts (hunk 10 above)
Comment (joshuatf) [reply plugins/woocommerce-admin/client/products/sections/test/utils.ts]: This test is no longer needed.

File: plugins/woocommerce-admin/client/products/sections/test/utils.ts (hunk 11 above)
Comment (joshuatf) [reply plugins/woocommerce-admin/client/products/sections/test/utils.ts]: This test is no longer needed.

**Comment by joshuatf** [comment]
@mattsherman @mdperez86 This PR has been updated to remove any validation/sanitization logic around SKUs.  I addressed a couple of @mattsherman's feedback requests, though a follow-up to improve form context callbacks may be a nice addition in the future.

File: plugins/woocommerce-admin/client/products/sections/product-inventory-section.tsx [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+
+	const setSkuIfEmpty = () => {
+		if ( values.sku || ! values.name?.length ) {
+			return;
+		}
+		setValue( 'sku', cleanForSlug( values.name ) );
+	};
+
+	const isNameBlurred = ! prevTouchedName && touched.name;
+	useEffect( setSkuIfEmpty, [ isNameBlurred ] );
```
Comment (mattsherman) [inline plugins/woocommerce-admin/client/products/sections/product-inventory-section.tsx]: This does not handle the scenario where the user:

1. Focuses on empty name field
2. Blurs to another field (without first entering anything in name field)
3. Focuses on empty name field
4. Enters name
5. Blurs to another field

Perhaps it would be better to go back to our original implementation, which tied to the `onBlur` of the field itself. I don't love that, since it couples the two UI implementations more than ideal, but it does work better. 

Perhaps the more ideal solution needs to be addressed in the future. Maybe we need a way to hook into focus change at the Form level, so we (and extensions) could hook into whether the focus changed from one component to another at the Form level.

Or, if you changed the implementation to always update the SKU on the fly as the name value changed, keystroke by keystroke, as long as it was a new product and the user hadn't manually modified the SKU. That would be pretty slick, as long as it wasn't a performance issue.

File: plugins/woocommerce-admin/client/products/sections/product-inventory-section.tsx (hunk 7 above)
Comment (mattsherman) [reply plugins/woocommerce-admin/client/products/sections/product-inventory-section.tsx:38]: I don't love clean up issues, but perhaps it would be best to address Tracks naming across the board in a follow up clean up issue, so we can get this functionality merged to unblock further MVP work.

As you mention, you were just following existing naming conventions on the new product experience. And, those existing conventions are not in line with best practices/guidelines. /cc @AnnaMag @pmcpinto

File: plugins/woocommerce-admin/client/products/sections/product-details-section.tsx [hunk 14]
```diff
@@ -24,9 +24,9 @@ import { recordEvent } from '@woocommerce/tracks';
  * Internal dependencies
  */
 import './product-details-section.scss';
+import { getCheckboxProps, getTextControlProps } from './utils';
```
Comment (mattsherman) [inline plugins/woocommerce-admin/client/products/sections/product-details-section.tsx:27]: Just curious why this changed... was is auto-changed by prettier/linter, or did you manually change it? Do we have guidelines around import ordering?

**Review by mattsherman — CHANGES_REQUESTED** [review]
Thanks for trying to improve the onBlur handling, @joshuatf! Unfortunately, I believe that your original implementation, while coupling the different UI sections less-than-ideally, functioned better. I left a comment suggesting we either go back to that implementation or change to an implementation where the SKU is updated on-the-fly as the name field changes.

File: plugins/woocommerce-admin/client/products/sections/product-inventory-section.tsx (hunk 13 above)
Comment (joshuatf) [reply plugins/woocommerce-admin/client/products/sections/product-inventory-section.tsx]: This was definitely a hacky solution at best.  I agree with moving back to the `onBlur`.  I'll open an issue to introduce a state handler to form context so we can detect various state changes to the form outside of specific components.

I've reverted this in https://github.com/woocommerce/woocommerce/pull/34978/commits/60da981d68f0ff9ff82742e2725d68c21b0423bc

File: plugins/woocommerce-admin/client/products/sections/product-details-section.tsx (hunk 14 above)
Comment (joshuatf) [reply plugins/woocommerce-admin/client/products/sections/product-details-section.tsx:27]: >was is auto-changed by prettier/linter, or did you manually change it?

I probably manually changed at this some point since I touched on the `utils` file at one point previously.  I did not opt to reorder all imports since that might litter the commit and make it harder to review what was changed.  But in retrospect, I think reordering all of them might be best otherwise it likely won't get updated in the future.

> Do we have guidelines around import ordering?

Yes, we are supposed to order alphabetically, both within the import destructure and the list itself.  For example:

```
import A from 'package-a';
import { B, D } from 'package-b';
import { C } from 'package-c';
```

AFAIK we don't have rules around unnamed imports, so scss files are often left at the top.

I don't see most Mothra or Ghidorah members following this, most likely because we don't linting rules around this.  I think it be worth surfacing these via p2 and introducing some linting rules to get everyone on the same page.

**Comment by joshuatf** [comment]
Thanks for the review @mattsherman!  I reverted the to the original `onBlur` method and this PR is ready for another review.

File: plugins/woocommerce-admin/client/products/sections/product-details-section.tsx (hunk 14 above)
Comment (mattsherman) [reply plugins/woocommerce-admin/client/products/sections/product-details-section.tsx:27]: Interested... did not know we had a guideline around this!

Comment (mattsherman) [reply plugins/woocommerce-admin/client/products/sections/product-details-section.tsx:27]: A linting rule would be excellent in the future.

**Review by mattsherman — DISMISSED** [review]
Nice work on this @joshuatf ! LGTM!

I have tested the latest, and it works as expected, with one ambiguous case. The original acceptance criteria didn't detail this, and I can see reasoning behind either behavior, so I'm going to approve this and we can follow up with a future issue if need be...

Scenario:

1. User enters in a name (SKU updates automatically)
2. User clears the SKU
3. User modifies the name (SKU updates automatically)

It is unclear to me what the "best" behavior is in this scenario... should the SKU stay blank because the user manually cleared it, or should it be auto-filled in, since it was blank? It's easy to clear out the SKU again, so I think that how you have it implemented is the preferred way, but it did not meet my expectations when testing, so I'm raising it for discussion.

/cc @jarekmorawski

**Comment by mattsherman** [comment]
@joshuatf Btw... I see the failing "Run tests against PR / Runs E2E tests", but the failures appear completely unrelated, so I think it is safe to bypass branch protections.

**Comment by joshuatf** [comment]
Thanks for the quick review, @mattsherman!

>It's easy to clear out the SKU again, so I think that how you have it implemented is the preferred way

I think so as well.  Worst case scenario, are there any detrimental effects to having a SKU filled in?  Or reasons a user might want a blank SKU?

@jarekmorawski if you have an opinion on this, please let us know and we'll follow up with another issue.

File: plugins/woocommerce-admin/client/products/sections/product-inventory-section.tsx (hunk 7 above)
Comment (AnnaMag) [reply plugins/woocommerce-admin/client/products/sections/product-inventory-section.tsx:38]: I do not see any existing wcadmin events that could be directly copied. I believe the pattern could be: `product_inventory_manage_settings_clicked`
(e.g. following the mobile event `product_detail_view_inventory_settings_tapped`).

**Review by mattsherman — APPROVED** [review]
Changes still look good to me.

**Comment by jarekmorawski** [comment]
> Worst case scenario, are there any detrimental effects to having a SKU filled in?

I don't think so. I'd go further and say that it's only beneficial since both WooCommerce and many inventory management systems that the merchants use treat the SKU as a product UID. It's perfectly fine to have it filled in again when the name changes.
